### ISSUE TITLE
SBV support for finite maps

### ISSUE THREAD

**Issue by andrew-wja** [LeventErkok/sbv#530]
Hello,

I noticed that SBV doesn't have a datatype like

```
SBV a, SBV b => SBV (Map a b)
```

so I went looking for such a datatype in SMT-lib and discovered that SMT-lib doesn't have such a datatype either (just arrays). I see that SBV does have a few datatypes which are built "on top" of SMT-lib rather than directly translated to an equivalent SMT-lib datatype.

Would you consider adding `SBV (Map a b)`? Or a pull request adding such a datatype?

I guess the underlying question here is how far down the path of building things on top of SMT-lib do you consider appropriate for SBV? For datatypes like `Maybe a` and `Either a b`, I can see the argument that they are both a) simple enough and b) so widely useful, that adding them feels necessary, while adding `Map a b` might not feel so necessary.

Right now, where I have models that logically contain structures that I'd like to describe with maps, I'm directly using Sets of tuples `(a, b)`, but that's a little unwieldy to work with. 

It would be very nice to have (like `SList` and `SSet`) the same typename and interface as the Haskell datatype.

**Comment by LeventErkok** [comment]
PR's are always welcome.

The problem with a map like data type is that it is really hard to implement it correctly and also make it efficient:

- If you require keys to be concrete, then you can just use Map a (SBV b); there's no need for symbolic maps.

- If you allow keys to be symbolic, which is obviously the more interesting case, then it is almost impossible to give a definition that will run "efficiently." You'll end up storing it as an array internally (most likely), and each symbolic-lookup will generate a lot of constraints.

This doesn't mean they wouldn't be useful; just a warning that it's really hard to get them working "efficiently." For instance, when I implemented `SFunArr` (which is supposed to be faster than internal SMT-arrays) I had to put in explicit caching to get reasonable performance. See #402 and #396 for some of the thorny issues that come up with such data structures. In particular, you want to make sure if you insert and element and just look it up, you really get that element very quickly. Think about what that would mean if the address is symbolic. Also, insert-with-merge like functions are necessarily linear.

Long story short: Please go ahead and give it a shot. But do include lots of tests to make sure performance is reasonable. A PR would be most welcome when you've something that suits your needs!

**Comment by andrew-wja** [comment]
Thanks for the in-depth explanation! I see now why they are not present.

I hadn't considered trying to move things that I'm doing out of the Symbolic domain partially like you suggest (concrete keys, symbolic values), but I have actually had great success wrapping the `Symbolic` monad transformer in a `StateT (Map a (SBV b))`, so my immediate problem is solved. It seems like efficient symbolic maps would be an interesting research project though, I will keep it in mind.

**Comment by LeventErkok** [comment]
@andrew-wja I'm going to close this ticket as there isn't any actionable item for SBV at this point. Feel free to reopen (or open a new issue) if there're other questions or designs you want to explore.
